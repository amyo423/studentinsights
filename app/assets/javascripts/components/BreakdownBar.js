import React from 'react';

// Visual component showing a horizontal bar broken down into a few colors
// showing percent of students of different attributes (eg, DIBELS scores, gender identity).
export default class BreakdownBar extends React.Component {
  totalCount() {
    const {items} = this.props;
    return items.reduce((sum, item) => sum + item.width, 0);
  }

  scale(count) {
    const totalCount = this.totalCount();
    return `${Math.ceil(100 * count / totalCount)}%`;
  }

  render() {
    const {items,height, style = {}} = this.props;    
    return (
      <div className="GenderBreakdownBar" style={{height, ...style}}>
        <div style={{position: 'relative', width: '100%', height}}>
          {items.map(item => this.renderBarAndLabel(item))}
        </div>
      </div>
    );
  }


  renderBarAndLabel({key, left, width, color}) {
    const {height, labelTop} = this.props;
    const opacity = 0.5;
    const fontSize = 10;

    if (width === 0) return;
    return (
      <div key={key}>
        <div style={{
          opacity,
          fontSize,
          position: 'absolute',
          background: color,
          left: this.scale(left),
          width: this.scale(width),
          height
        }}>{'\u00A0'}</div>
        {width > 0 && 
          <div style={{
            opacity,
            fontSize,
            color,
            position: 'absolute',
            textAlign: 'right',
            left: this.scale(left),
            width: this.scale(width),
            top: labelTop,
            paddingRight: 1
          }}>{width}</div>}
      </div>
    );
  }
}

BreakdownBar.propTypes = {
  items: React.PropTypes.arrayOf(React.PropTypes.shape({
    left: React.PropTypes.number.isRequired,
    width: React.PropTypes.number.isRequired,
    color: React.PropTypes.string.isRequired
  })).isRequired,
  height: React.PropTypes.number.isRequired,
  labelTop: React.PropTypes.number.isRequired,
  style: React.PropTypes.object
};
