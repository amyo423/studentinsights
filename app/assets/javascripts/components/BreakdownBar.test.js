import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import BreakdownBar from './BreakdownBar';


export function testProps(props) {
  return {
    items: [
      { left: 0, width: 7, color: 'green', key: 'core' },
      { left: 7, width: 2, color: 'orange', key: 'strategic' },
      { left: 7 + 2, width: 3, color: 'red', key: 'intensive' }
    ],
    height: 50,
    labelTop: 55,
    ...props
  };
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  const props = testProps();
  ReactDOM.render(<BreakdownBar {...props} />, el);
});

it('snapshots', () => {
  const props = testProps();
  const tree = renderer
    .create(<BreakdownBar {...props} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});