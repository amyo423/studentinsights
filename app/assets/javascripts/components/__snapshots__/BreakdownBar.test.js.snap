// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="BreakdownBar"
  style={
    Object {
      "height": 50,
    }
  }
>
  <div
    style={
      Object {
        "height": 50,
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div>
      <div
        style={
          Object {
            "background": "green",
            "height": 50,
            "left": "0%",
            "position": "absolute",
            "width": "59%",
          }
        }
      >
         
      </div>
      <div
        style={
          Object {
            "color": "green",
            "left": "0%",
            "paddingRight": 1,
            "position": "absolute",
            "textAlign": "right",
            "top": 55,
            "width": "59%",
          }
        }
      >
        7
      </div>
    </div>
    <div>
      <div
        style={
          Object {
            "background": "orange",
            "height": 50,
            "left": "59%",
            "position": "absolute",
            "width": "17%",
          }
        }
      >
         
      </div>
      <div
        style={
          Object {
            "color": "orange",
            "left": "59%",
            "paddingRight": 1,
            "position": "absolute",
            "textAlign": "right",
            "top": 55,
            "width": "17%",
          }
        }
      >
        2
      </div>
    </div>
    <div>
      <div
        style={
          Object {
            "background": "red",
            "height": 50,
            "left": "75%",
            "position": "absolute",
            "width": "25%",
          }
        }
      >
         
      </div>
      <div
        style={
          Object {
            "color": "red",
            "left": "75%",
            "paddingRight": 1,
            "position": "absolute",
            "textAlign": "right",
            "top": 55,
            "width": "25%",
          }
        }
      >
        3
      </div>
    </div>
  </div>
</div>
`;
